core/src/main/java/io/questdb/griffin/engine/join/AsOfJoinMemoizedRecordCursorFactory.java

```diff
@@ -0,0 +1,355 @@
+                    }
+                    if (rememberedRowId >= 0) {
+                        // We saw this symbol at rememberedRowId. We can now reuse it and complete the search.
+                        if (isSlaveWithinToleranceInterval(masterTimestamp, validityPeriodStart)) {
+                            record.hasSlave(true);
+                            slaveTimeFrameCursor.recordAt(slaveRecB, rememberedRowId);
+                        } else {
+                            record.hasSlave(false);
+                        }
+                        break;
+                    } else {
+                        // We remembered a period within which the symbol doesn't occur. Jump over the entire
+                        // period and continue searching, unless the remembered period extends beyond the
+                        // tolerance interval. If we have just extended the validity period, it definitely
+                        // extends beyond the tolerance interval.
+                        if (didExtendValidityPeriod || !isSlaveWithinToleranceInterval(masterTimestamp, validityPeriodStart)) {
+                            record.hasSlave(false);
+                            break;
+                        }
+                        didJumpOverValidityPeriod = true;
+                        rowId = -rememberedRowId - 1;
+                        int frameIndex = Rows.toPartitionIndex(rowId);
+                        slaveTimeFrameCursor.jumpTo(frameIndex);
+                        slaveTimeFrameCursor.open();
+                        slaveTimeFrameCursor.recordAt(slaveRecB, rowId);
+                        frameRowLo = Rows.toRowID(frameIndex, slaveTimeFrame.getRowLo());
+                        slaveTimestamp = scaleTimestamp(slaveRecB.getTimestamp(slaveTimestampIndex), slaveTimestampScale);
+                    }
+                }
+                if (!isSlaveWithinToleranceInterval(masterTimestamp, slaveTimestamp)) {
+                    // We have been searching outside the remembered period and are past the tolerance interval.
+                    // Stop and report no match.
+                    record.hasSlave(false);
+                    // memorize that we didn't find the matching symbol by saving rowId as (-rowId - 1)
+                    memorizeSymbolLocation(masterTimestamp, slaveTimestamp, slaveSymbolKey, -rowId - 1, false);
+                    break;
+                }
+                int thisSymbolKey = slaveRecB.getInt(slaveSymbolColumnIndex);
+                if (thisSymbolKey == slaveSymbolKey) {
+                    record.hasSlave(true);
+                    // We found the symbol that we don't remember. Memorize it now.
+                    memorizeSymbolLocation(masterTimestamp, slaveTimestamp, slaveSymbolKey, rowId, false);
+                    break;
+                } else {
+                    // This isn't the symbol we're looking for, but memorize it anyway in the hope that some future
```

Comment:
Shouldn't be update the symbol values in both cases: when it's an unseen key and when it's an already seen one? In case of sparse timestamps, this could save us some jumps across the page frames.

Reply:
After more thinking, I recalled this problem:

As soon as we lose the invariant that this is the only symbol between `masterTimestamp` and the place where we found it, we're lost and don't know the validity period of the symbol. We have this invariant when the symbol is the one we're looking for, but for these "drive-by" symbols, we don't know if we see it for the first time in the scanned region.

This forces the logic to stay the way it is.